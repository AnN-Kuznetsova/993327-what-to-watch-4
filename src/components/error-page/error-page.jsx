import React from "react";
import PropTypes from "prop-types";

import {Header} from "../header/header";
import {Footer} from "../footer/footer";


const getErrorMessage = (dataError) => {
  if (dataError) {
    return `The request failed`;
  }

  return null;
};


export const ErrorPage = (props) => {
  const {dataError} = props;

  return (
    <React.Fragment>
      <section
        className="movie-card"
        style={{backgroundColor: `#180202`}}
      >
        <Header />

        <div className="movie-card__wrap">
          <div className="movie-card__info">
            <div className="movie-card__desc">
              <h2 className="movie-card__title">{getErrorMessage(dataError)}</h2>
            </div>
          </div>
        </div>
      </section>

      <div className="page-content">
        <Footer />
      </div>
    </React.Fragment>
  );
};


ErrorPage.propTypes = {
  dataError: PropTypes.bool.isRequired,
};
