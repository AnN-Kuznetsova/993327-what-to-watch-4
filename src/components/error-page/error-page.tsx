import * as React from "react";

import {Error} from "../../api";
import {Footer} from "../footer/footer";
import {Header} from "../header/header";


interface Props {
  dataError?: object;
}


const getErrorMessage = (dataError) => {
  if (dataError) {
    switch (true) {
      case dataError.status === Error.PAGE_NOT_FOUND:
        return `Page not found.`;

      default:
        return `The request failed`;
    }
  }

  return null;
};


const ErrorPage: React.FunctionComponent<Props> = (props: Props) => {
  const {dataError} = props;

  return (
    <React.Fragment>
      <section
        className="movie-card"
        style={{backgroundColor: `#180202`}}
      >
        <Header />

        <div className="movie-card__wrap">
          <div className="movie-card__info">
            <div className="movie-card__desc">
              <h2 className="movie-card__title">{getErrorMessage(dataError)}</h2>
            </div>
          </div>
        </div>
      </section>

      <div className="page-content">
        <Footer />
      </div>
    </React.Fragment>
  );
};


export {
  ErrorPage,
};
