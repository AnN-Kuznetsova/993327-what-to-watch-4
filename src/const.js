const NUMBER_OF_CARDS_IN_CATALOG_AT_STARTUP = 8;
const NUMBER_OF_CARDS_TO_INCREMENT = 8;
const MAX_FILTER_ELEMENTS_COUNT = 10;
const NUMBER_OF_SIMILAR_FILMS = 4;
const VISIBLE_PARTICIPANTS_COUNT = 4;
const DELAY_PLAYBACK_PREVIEW = 1000;
const MIN_REVIEW_TEXT_LENGTH = 50;
const MAX_REVIEW_TEXT_LENGTH = 400;
const RATING_RANGE = 5;
const ERROR_COLOR = `#a8421e`;

const BASENAME_FOR_DEPLOY = `/htmlacademy-what-to-watch`;

const AppRoute = {
  MAIN: BASENAME_FOR_DEPLOY + `/`,
  SIGN_IN: BASENAME_FOR_DEPLOY + `/login`,
  MY_LIST: BASENAME_FOR_DEPLOY + `/myList`,
  FILM: BASENAME_FOR_DEPLOY + `/films/:id`,
  ADD_REVIEW: BASENAME_FOR_DEPLOY + `/films/:id/review`,
  PLAYER: BASENAME_FOR_DEPLOY + `/player/:id`,
};


export {
  DELAY_PLAYBACK_PREVIEW,
  ERROR_COLOR,
  MAX_FILTER_ELEMENTS_COUNT,
  MAX_REVIEW_TEXT_LENGTH,
  MIN_REVIEW_TEXT_LENGTH,
  NUMBER_OF_CARDS_IN_CATALOG_AT_STARTUP,
  NUMBER_OF_CARDS_TO_INCREMENT,
  NUMBER_OF_SIMILAR_FILMS,
  RATING_RANGE,
  VISIBLE_PARTICIPANTS_COUNT,
  AppRoute,
  BASENAME_FOR_DEPLOY,
};
