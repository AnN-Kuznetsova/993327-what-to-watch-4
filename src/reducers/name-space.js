export const NameSpace = {
  DATA: `DATA`,
  APPLICATION: `APPLICATION`,
  USER: `USER`,
};
