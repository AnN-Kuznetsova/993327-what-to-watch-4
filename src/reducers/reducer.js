import {
  NUMBER_OF_CARDS_IN_CATALOG_AT_STARTUP,
  NUMBER_OF_CARDS_TO_INCREMENT,
  PageType,
  FilterType,
} from "../const";
import {
  extend,
  getFilteredMovies,
} from "../utils/utils";

import {movies} from "../mocks/movies";
import {promoMovie} from "../mocks/promo-movie";


const initialState = {
  movies,
  genre: `All genres`,
  movieList: movies,
  visibleMoviesCount: movies.length <= NUMBER_OF_CARDS_IN_CATALOG_AT_STARTUP ? movies.length :
    NUMBER_OF_CARDS_IN_CATALOG_AT_STARTUP,
  activeMovie: promoMovie,
  activePage: PageType.MAIN,
  prevPage: PageType.MAIN,
};


const ActionType = {
  CHANGE_GENRE: `CHANGE_GENRE`,
  GET_MOVIES: `GET_MOVIES`,
  CHANGE_ACTIVE_MOVIE: `CHANGE_ACTIVE_MOVIE`,
  CHANGE_ACTIVE_PAGE: `CHANGE_ACTIVE_PAGE`,
  INCREMENT_VISIBLE_MOVIES_COUNT: `INCREMENT_VISIBLE_MOVIES_COUNT`,
  RESET_VISIBLE_MOVIES_COUNT: `RESET_VISIBLE_MOVIES_COUNT`,
};

const ActionCreator = {
  changeGenre: (value) => ({
    type: ActionType.CHANGE_GENRE,
    payload: value,
  }),
  getMovies: (moviesCount = null) => ({
    type: ActionType.GET_MOVIES,
    payload: moviesCount,
  }),
  changeActiveMovie: (movie) => ({
    type: ActionType.CHANGE_ACTIVE_MOVIE,
    payload: movie,
  }),
  changeActivePage: (page) => ({
    type: ActionType.CHANGE_ACTIVE_PAGE,
    payload: page,
  }),
  incrementVisibleMoviesCount: () => ({
    type: ActionType.INCREMENT_VISIBLE_MOVIES_COUNT,
    payload: NUMBER_OF_CARDS_TO_INCREMENT,
  }),
  resetVisibleMoviesCount: () => ({
    type: ActionType.RESET_VISIBLE_MOVIES_COUNT,
    payload: null,
  }),
};


const reducer = (state = initialState, action) => {
  switch (action.type) {
    case ActionType.CHANGE_GENRE:
      return extend(state, {
        genre: action.payload,
      });

    case ActionType.GET_MOVIES:
      return extend(state, {
        movieList: getFilteredMovies(state.movies, FilterType.GENRE, state.genre, state.activeMovie, action.payload),
      });

    case ActionType.CHANGE_ACTIVE_MOVIE:
      return extend(state, {
        activeMovie: action.payload,
      });

    case ActionType.CHANGE_ACTIVE_PAGE:
      if (state.activePage === action.payload) {
        return state;
      }
      return extend(state, {
        prevPage: state.activePage,
        activePage: action.payload,
      });

    case ActionType.INCREMENT_VISIBLE_MOVIES_COUNT:
      return extend(state, {
        visibleMoviesCount: state.visibleMoviesCount + action.payload,
      });

    case ActionType.RESET_VISIBLE_MOVIES_COUNT:
      return extend(state, {
        visibleMoviesCount: initialState.visibleMoviesCount,
      });

    default:
      return state;
  }
};


export {
  ActionCreator,
  ActionType,
  reducer,
};
