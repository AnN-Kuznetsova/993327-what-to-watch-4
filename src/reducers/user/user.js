import {extend} from "../../utils/utils";


const AuthorizationStatus = {
  AUTH: `AUTH`,
  NO_AUTH: `NO_AUTH`,
};

const initialState = {
  authorizationStatus: AuthorizationStatus.NO_AUTH,
  loginError: null,
};

const ActionType = {
  REQUIRED_AUTHORIZATION: `REQUIRED_AUTHORIZATION`,
  SET_LOGIN_ERROR: `SET_LOGIN_ERROR`,
};

const ActionCreator = {
  requireAuthorization: (status) => ({
    type: ActionType.REQUIRED_AUTHORIZATION,
    payload: status,
  }),
  setLoginError: (error) => ({
    type: ActionType.SET_LOGIN_ERROR,
    payload: error,
  }),
};

const Operation = {
  checkAuth: () => (dispatch, getState, api) => {
    return api.get(`/login`)
      .then(() => {
        dispatch(ActionCreator.requireAuthorization(AuthorizationStatus.AUTH));
      })
      .catch((err) => {
        throw err;
      });
  },

  login: (authData) => (dispatch, getState, api) => {
    return api.post(`/login`, {
      email: authData.email,
      password: authData.password,
    })
    .then(() => {
      dispatch(ActionCreator.requireAuthorization(AuthorizationStatus.AUTH));
      dispatch(ActionCreator.setLoginError(null));
    })
    .catch((error) => {
      dispatch(ActionCreator.setLoginError(error));
    });
  },
};

const reducer = (state = initialState, action) => {
  switch (action.type) {
    case ActionType.REQUIRED_AUTHORIZATION:
      return extend(state, {
        authorizationStatus: action.payload,
      });

    case ActionType.SET_LOGIN_ERROR:
      return extend(state, {
        loginError: action.payload
      });

    default:
      return state;
  }
};


export {
  ActionCreator,
  ActionType,
  AuthorizationStatus,
  Operation,
  reducer,
};
