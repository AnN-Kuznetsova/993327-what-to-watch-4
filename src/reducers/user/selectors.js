import {NameSpace} from "../name-space.js";


const NAME_SPASE = NameSpace.USER;


export const getAuthorizationStatus = (state) => {
  return state[NAME_SPASE].authorizationStatus;
};
