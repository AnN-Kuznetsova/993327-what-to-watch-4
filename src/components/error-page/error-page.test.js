import React from "react";
import configureStore from "redux-mock-store";
import renderer from "react-test-renderer";
import {Provider} from "react-redux";

import {AuthorizationStatus} from "../../reducers/user/user";
import {ErrorPage} from "./error-page";
import {NameSpace} from "../../reducers/name-space";
import {PageType} from "../../const";


const mockStore = configureStore([]);

const store = mockStore({
  [NameSpace.USER]: {
    authorizationStatus: AuthorizationStatus.NO_AUTH,
  },
  [NameSpace.APPLICATION]: {
    activePage: PageType.ERROR,
  },
});


describe(`Render ErrorPage`, () => {
  it(`ErrorPage should match with snapshot when dataError is null`, () => {
    const errorPageSnapshot = renderer.create(
        <Provider store={store} >
          <ErrorPage dataError={null} />
        </Provider>
    ).toJSON();

    expect(errorPageSnapshot).toMatchSnapshot();
  });


  it(`ErrorPage should match with snapshot when dataError is true`, () => {
    const errorPageSnapshot = renderer.create(
        <Provider store={store} >
          <ErrorPage dataError={{request: true}} />
        </Provider>
    ).toJSON();

    expect(errorPageSnapshot).toMatchSnapshot();
  });
});
