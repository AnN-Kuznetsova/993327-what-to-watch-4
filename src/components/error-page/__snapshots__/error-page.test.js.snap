// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Render ErrorPage ErrorPage should match with snapshot when dataError is null 1`] = `
Array [
  <section
    className="movie-card"
    style={
      Object {
        "backgroundColor": "#180202",
      }
    }
  >
    <h1
      className="visually-hidden"
    >
      WTW
    </h1>
    <header
      className="page-header false false"
    >
      <div
        className="logo"
      >
        <a
          className="logo__link false"
          href="/"
        >
          <span
            className="logo__letter logo__letter--1"
          >
            W
          </span>
          <span
            className="logo__letter logo__letter--2"
          >
            T
          </span>
          <span
            className="logo__letter logo__letter--3"
          >
            W
          </span>
        </a>
      </div>
    </header>
    <div
      className="movie-card__wrap"
    >
      <div
        className="movie-card__info"
      >
        <div
          className="movie-card__desc"
        >
          <h2
            className="movie-card__title"
          />
        </div>
      </div>
    </div>
  </section>,
  <div
    className="page-content"
  >
    <footer
      className="page-footer"
    >
      <div
        className="logo"
      >
        <a
          className="logo__link logo__link--light"
          href="/"
        >
          <span
            className="logo__letter logo__letter--1"
          >
            W
          </span>
          <span
            className="logo__letter logo__letter--2"
          >
            T
          </span>
          <span
            className="logo__letter logo__letter--3"
          >
            W
          </span>
        </a>
      </div>
      <div
        className="copyright"
      >
        <p>
          © 2019 What to watch Ltd.
        </p>
      </div>
    </footer>
  </div>,
]
`;

exports[`Render ErrorPage ErrorPage should match with snapshot when dataError is true 1`] = `
Array [
  <section
    className="movie-card"
    style={
      Object {
        "backgroundColor": "#180202",
      }
    }
  >
    <h1
      className="visually-hidden"
    >
      WTW
    </h1>
    <header
      className="page-header false false"
    >
      <div
        className="logo"
      >
        <a
          className="logo__link false"
          href="/"
        >
          <span
            className="logo__letter logo__letter--1"
          >
            W
          </span>
          <span
            className="logo__letter logo__letter--2"
          >
            T
          </span>
          <span
            className="logo__letter logo__letter--3"
          >
            W
          </span>
        </a>
      </div>
    </header>
    <div
      className="movie-card__wrap"
    >
      <div
        className="movie-card__info"
      >
        <div
          className="movie-card__desc"
        >
          <h2
            className="movie-card__title"
          >
            The request failed
          </h2>
        </div>
      </div>
    </div>
  </section>,
  <div
    className="page-content"
  >
    <footer
      className="page-footer"
    >
      <div
        className="logo"
      >
        <a
          className="logo__link logo__link--light"
          href="/"
        >
          <span
            className="logo__letter logo__letter--1"
          >
            W
          </span>
          <span
            className="logo__letter logo__letter--2"
          >
            T
          </span>
          <span
            className="logo__letter logo__letter--3"
          >
            W
          </span>
        </a>
      </div>
      <div
        className="copyright"
      >
        <p>
          © 2019 What to watch Ltd.
        </p>
      </div>
    </footer>
  </div>,
]
`;
